src/metabase/task/sync_databases.clj

```diff
@@ -238,63 +241,71 @@
-   (triggers/with-identity (trigger-key database task-info))
-   (triggers/using-job-data {"db-id" (u/the-id database)})
-   (triggers/for-job (job-key task-info))
-   (triggers/start-now)
-   (triggers/with-schedule
-     (cron/schedule
-      (cron/cron-schedule (cron-schedule database task-info))
-      ;; if we miss a sync for one reason or another (such as system being down) do not try to run the sync again.
-      ;; Just wait until the next sync cycle.
-      ;;
-      ;; See https://www.nurkiewicz.com/2012/04/quartz-scheduler-misfire-instructions.html for more info
-      (cron/with-misfire-handling-instruction-do-nothing)))))
+  (when-let [task-schedule (cron-schedule database task-info)]
+    (triggers/build
+     (triggers/with-description (trigger-description database task-info))
+     (triggers/with-identity (trigger-key database task-info))
+     (triggers/using-job-data {"db-id" (u/the-id database)})
+     (triggers/for-job (job-key task-info))
+     (triggers/start-now)
+     (triggers/with-schedule
+       (cron/schedule
+        (cron/cron-schedule task-schedule)
+        ;; if we miss a sync for one reason or another (such as system being down) do not try to run the sync again.
+        ;; Just wait until the next sync cycle.
+        ;;
+        ;; See https://www.nurkiewicz.com/2012/04/quartz-scheduler-misfire-instructions.html for more info
+        (cron/with-misfire-handling-instruction-do-nothing))))))
+
+(defn- update-db-trigger-if-needed!
+  "Replace or remove the existing trigger if the schedule changes, do nothing if schedule is the same."
+  [database task-info]
+  (let [job                                 (task/job-info (job-key task-info))
+        new-trigger                         (trigger database task-info)
+        ;; there should be only one schedule per task per DB
+        existing-trigger-with-same-schedule (when new-trigger
+                                              (some #(when (and (= (:key %) (.. new-trigger getKey getName))
+                                                                (= (:schedule %) (cron-schedule database task-info)))
+                                                       %)
+                                                    (:triggers job)))]
+    (cond
+     ;; no new schedule
+     ;; delete the existing trigger
+     (nil? new-trigger)
+     (do
+      (log/infof "Trigger for \"%s\" of Database \"%s\" has been removed."
+                 (:name task-info)
+                 (:name database))
+      (delete-trigger! database task-info))
+
+     ;; need to recreate the new trigger
+     (and (some? new-trigger)
+          (nil? existing-trigger-with-same-schedule))
+     (do
+      (log/infof "Trigger for \"%s\" of Database \"%s\" is now \"%s\""
+                 (:name task-info)
+                 (:name database)
+                 (cron-schedule database task-info))
+      ;; delete all existing triggers just to be sure
```

Comment:
It could be nice to log if there were duplicates that we implicitly removed.

Reply:
trigger key has a unique constraint, and our naming for the trigger is unique per task per DB so it should be no duplicated triggers.

we will only delete the existing one.

tried to make it a bit nicer here(8fbfcf71fe) wdyt?

Reply:
Yeah that's much clearer!